internal/syncronizedmap/syncronizedmap.go

```diff
@@ -0,0 +1,67 @@
+package syncronizedmap
+
+import (
+	"fmt"
+	"sync"
+)
+
+//go:generate mockgen -source=syncronizedmap.go -package=syncronizedmap -destination=mock_syncronizedmap.go
+
+type KernelOsDtkMapping interface {
+	NewNodeInfo(kernelVersion, osImage string)
+	NewImageStreamInfo(osImage, dtkImage string)
+	GetImage(kernelVersion string) (string, error)
+}
+
+//FIXME: add UT
+
+type kernelOsDtkMapping struct {
+	kernelToOsMutex *sync.RWMutex
+	osToDtkMutext   *sync.RWMutex
+	kernelToOs      map[string]string
+	osToDtk         map[string]string
+}
+
+func NewKernelOsDtkMapping() KernelOsDtkMapping {
+
+	return &kernelOsDtkMapping{
+		kernelToOsMutex: &sync.RWMutex{},
+		osToDtkMutext:   &sync.RWMutex{},
+		kernelToOs:      map[string]string{},
+		osToDtk:         map[string]string{},
+	}
+}
+
+func (skom *kernelOsDtkMapping) NewNodeInfo(kernelVersion, osImage string) {
+
+	skom.kernelToOsMutex.Lock()
+	defer skom.kernelToOsMutex.Unlock()
+
+	skom.kernelToOs[kernelVersion] = osImage
+}
+
+func (skom *kernelOsDtkMapping) NewImageStreamInfo(osImage, dtkImage string) {
+
+	skom.osToDtkMutext.Lock()
+	defer skom.osToDtkMutext.Unlock()
+
+	skom.osToDtk[osImage] = dtkImage
+}
+
+func (skom *kernelOsDtkMapping) GetImage(kernelVersion string) (string, error) {
+
+	skom.kernelToOsMutex.RLock()
+	skom.osToDtkMutext.RLock()
+	defer skom.kernelToOsMutex.RUnlock()
+	defer skom.osToDtkMutext.RUnlock()
+
+	osImage, ok := skom.kernelToOs[kernelVersion]
+	if !ok {
+		return "", fmt.Errorf("could not find kernel %s in kernel --> OS mapping", kernelVersion)
+	}
+	dtk, ok := skom.osToDtk[osImage]
+	if !ok {
+		return "", fmt.Errorf("could not find osImage %s in OS --> DTK mapping", osImage)
+	}
+	return dtk, nil
```

Comment:
Let us lock the OS --> DTK lock as late as possible:
```suggestion
	skom.kernelToOsMutex.RLock()
	defer skom.kernelToOsMutex.RUnlock()

	osImage, ok := skom.kernelToOs[kernelVersion]
	if !ok {
		return "", fmt.Errorf("could not find kernel %s in kernel --> OS mapping", kernelVersion)
	}
	
	skom.osToDtkMutext.RLock()
	defer skom.osToDtkMutext.RUnlock()

	dtk, ok := skom.osToDtk[osImage]
	if !ok {
		return "", fmt.Errorf("could not find osImage %s in OS --> DTK mapping", osImage)
	}
	return dtk, nil
```
Ideally we would also release the kernel --> OS one earlier than at the end of the function, but that is error-prone since we want to immediately return an error if the kernel key does not exist.

Reply:
Don.e